### DISCUSSION TITLE
Chrome Extension: Can't Init the Web3Wallet SDK

### DISCUSSION THREAD

**Discussion by ManuelFerreras** [WalletConnect/walletconnect-monorepo#2986] [Developer Support] [answered]
Hi. I'm trying to init the Web3Wallet SDK following the instructions and the demo code. This is intended to be for a chrome extension.

My code is as follows:

`import { Core } from '@walletconnect/core'
import { buildApprovedNamespaces, getSdkError } from '@walletconnect/utils'
import { ErrorResponse, JsonRpcResponse } from '@walletconnect/jsonrpc-types'
import { ICore } from '@walletconnect/types'
import {
  IWeb3Wallet,
  Web3Wallet,
  Web3WalletTypes
} from '@walletconnect/web3wallet'

export let web3Wallet: IWeb3Wallet
export let core: ICore

export async function createWeb3Wallet() {
  console.debug('createWeb3Wallet')

  core = new Core({
    projectId: 'xxxxxxxxxxxxxxx'
  })

  web3Wallet = await Web3Wallet.init({
    core,
    metadata: {
      name: 'xxx',
      description: 'xxx',
      url: 'xxx',
      icons: ['xxx']
    }
  })
}`

and in the browser I get:

`
browser.js:202 {context: 'core'} {context: 'core/relayer'}context: "core/relayer"[[Prototype]]: Object TypeError: o.on is not a function
    at ws.ts:128:25
    at new Promise (<anonymous>)
    at e.value (ws.ts:119:12)
    at e.<anonymous> (ws.ts:58:16)
    at h (regeneratorRuntime.js:44:17)
    at Generator.<anonymous> (regeneratorRuntime.js:125:22)
    at Generator.next (regeneratorRuntime.js:69:21)
    at x (asyncToGenerator.js:3:20)
    at s (asyncToGenerator.js:22:9)
    at asyncToGenerator.js:27:7

ws.ts:128 Uncaught (in promise) TypeError: o.on is not a function
    at ws.ts:128:25
    at new Promise (<anonymous>)
    at e.value (ws.ts:119:12)
    at e.<anonymous> (ws.ts:58:16)
    at h (regeneratorRuntime.js:44:17)
    at Generator.<anonymous> (regeneratorRuntime.js:125:22)
    at Generator.next (regeneratorRuntime.js:69:21)
    at x (asyncToGenerator.js:3:20)
    at s (asyncToGenerator.js:22:9)
    at asyncToGenerator.js:27:7`

Any ideas on what I can do?

Also, I'm using the following libraries:
    "@walletconnect/client": "^1.8.0",
    "@walletconnect/core": "^2.8.6",
    "@walletconnect/jsonrpc-types": "^1.0.2",
    "@walletconnect/types": "^2.8.6",
    "@walletconnect/utils": "^2.7.4",
    "@walletconnect/web3wallet": "^1.8.5",

**Comment by boidushya** [comment]
@ignaciosantise Can you take a look at this?

**Comment by ignaciosantise** [comment]
hey @ManuelFerreras 👋 can you try removing `@walletconnect/client` dependency and check again? That's our v1 legacy client.

If the problem persists, can you provide a reproducible example? It can be a public repository, codesandbox

**Reply by ManuelFerreras** [reply]
hey! sorry for the delays. here is a codesandbox, but it throws a completely different error:

[here](https://codesandbox.io/p/github/codesandbox/codesandbox-template-vite-react/csb-2dqzgl/draft/fervent-chatterjee?file=%2Fsrc%2FApp.tsx%3A1%2C1&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljoq9zjq000b3b6t0wz4hrxi%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522panelType%2522%253A%2522TABS%2522%252C%2522id%2522%253A%2522cljoq9zjr000d3b6tzbj4jxz1%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522cljoq9zjq000b3b6t0wz4hrxi%2522%253A%257B%2522id%2522%253A%2522cljoq9zjq000b3b6t0wz4hrxi%2522%252C%2522activeTabId%2522%253A%2522cljor2846025c3b6st1i5qo99%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljoq9zjq000a3b6thut79zlm%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252FApp.tsx%2522%252C%2522id%2522%253A%2522cljoqa91z006w3b6t4julren4%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522cljoqgj8c00w93b6s7pi3jukv%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TERMINAL%2522%252C%2522shellId%2522%253A%2522cljoqgjwp00r6fogfda2o32xi%2522%257D%252C%257B%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252Fsrc%252Futils.ts%2522%252C%2522id%2522%253A%2522cljor2846025c3b6st1i5qo99%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%257D%252C%2522cljoq9zjr000d3b6tzbj4jxz1%2522%253A%257B%2522id%2522%253A%2522cljoq9zjr000d3b6tzbj4jxz1%2522%252C%2522activeTabId%2522%253A%2522cljoqeipe005z3b6sh8ak554s%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522cljoqeipe005z3b6sh8ak554s%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522port%2522%253A5173%252C%2522taskId%2522%253A%2522yarn%2520dev%2522%252C%2522path%2522%253A%2522%252F%2522%257D%252C%257B%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522yarn%2520dev%2522%252C%2522id%2522%253A%2522cljor0bhb01pz3b6stsczrljn%2522%252C%2522mode%2522%253A%2522permanent%2522%257D%255D%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D)

**Reply by ManuelFerreras** [reply]
@ignaciosantise sorry to bother you, but this is very urgent. are there any errors in the snippet I shared?

**Reply by ignaciosantise** [reply]
@ManuelFerreras the code seems OK, so i think there's an problem with codesandbox. Can you fork this repo and reproduce your error here? https://github.com/WalletConnect/web-examples/tree/main/wallets/react-web3wallet

**Reply by ManuelFerreras** [reply]
@ignaciosantise so I tried in that repo locally, and it worked. I also created a [brand new repo with next](https://github.com/ManuelFerreras/next-wcv2-playground), and it also worked. The problem seems to be related to our app being a Chrome extension and having the SDK running in the background script or something related. Do you know if there is any code base to check why I'm having this problem?
<img width="1006" alt="Screenshot 2023-07-05 at 15 25 33" src="https://github.com/WalletConnect/walletconnect-monorepo/assets/35044975/d66fe442-0386-4e49-8c00-5eb0e45695f8">

**Reply by ignaciosantise** [reply]
I'll check with the team if we have anything related to chrome extensions to help you

**Reply by bkrem** [reply]
@ManuelFerreras This seems to be a specific issue around either:
* usage of web3wallet inside a Chrome extension generally (I'm not aware of other reports atm but possible of course)
* the way you are processing the dependency for the extension/`background.js` use case

Could you provide us with a minimal repro of the concrete scenario you're in? We currently don't have internal samples for Chrome extensions.

**Reply by ManuelFerreras** [reply]
@bkrem I'll create a minimal example a send you the repo ASAP

**Reply by ManuelFerreras** [reply]
@bkrem @ignaciosantise here is a minimal implementation in the background service worker of a chrome extension made with vite:

https://github.com/ManuelFerreras/vite-wev2-playground-extension

to install it just run npm i in the root, then npm run dev, and a dist folder should be created. go to chrome extensions -> pack unloaded -> and select the dist folder.

if you open the inspector or the background service console, you will be able to see the error. it seems it has to do with the WS connection

**Reply by ManuelFerreras** [reply]
@bkrem this is the error:
` 'ws does not work in the browser. Browser clients must use the native WebSocket object'`

**Reply by bkrem** [reply, marked as answer]
@ManuelFerreras Thank you for the repro + instructions, that helped a lot in pinning down the root cause of the issue 🙏 

I've debugged what the issue is around the websocket resolution logic, and the following canary releases should unblock you until we can get this patched on `web3wallet@latest`:
```
    "@walletconnect/core": "2.9.0-e9ace057",
    "@walletconnect/types": "2.9.0-e9ace057",
    "@walletconnect/web3wallet": "1.8.6-e9ace057",
```

With the above deps your repro seems to initialise Web3Wallet correctly now.

Issue in a nutshell: The websocket resolution logic attempts to resolve `WebSocket` on `global` and `window` before defaulting to the Node `ws` implementation. But this turns out to be non-exhaustive since there _is_ a `WebSocket` built-in in the extension env (i.e. just `WebSocket` or `self.WebSocket`) but no `window` which it would normally resolve to in a browser env.

I'm creating a dedicated issue for this to track this fix until release 👍

**Reply by ManuelFerreras** [reply]
Hi @bkrem thank you very much for this! I'll try this right now and let you know if it works for me 🙂

**Comment by bkrem** [comment]
Tracking bug fix for this here: https://github.com/WalletConnect/walletconnect-utils/issues/116

Currently available patched canaries are:
```
    "@walletconnect/core": "2.9.0-e9ace057",
    "@walletconnect/types": "2.9.0-e9ace057",
    "@walletconnect/web3wallet": "1.8.6-e9ace057",
```

**Reply by ManuelFerreras** [reply]
@bkrem is there a way to know when this fix is going to be pushed to an official version?

**Reply by ManuelFerreras** [reply]
BTW, I'm reading the docs about submitting a Wallet to the WC Explorer, but all the links to test the wallet are returning a 404. They are all the links in this url:
https://docs.walletconnect.com/2.0/advanced/migration-from-v1.x/explorer-submission

**Reply by bkrem** [reply]
> @bkrem is there a way to know when this fix is going to be pushed to an official version?

Aiming for Monday 👍 

> BTW, I'm reading the docs about submitting a Wallet to the WC Explorer, but all the links to test the wallet are returning a 404. They are all the links in this url

Ah I see, the lab.web3modal.com URLs are stale there. Will let the team know so we can get these fixed. Thanks for reporting 🙏

**Reply by bkrem** [reply]
Links have been fixed 👍

**Reply by ManuelFerreras** [reply]
awesome, thanks!

**Reply by bkrem** [reply]
@ManuelFerreras This fix has now been shipped on the `latest` distro as https://github.com/WalletConnect/walletconnect-monorepo/releases/tag/2.9.0

Thanks again for your help on getting to the bottom of this via the repro 🙏

**Reply by ManuelFerreras** [reply]
awesome @bkrem . thank you for all this!

**Reply by ManuelFerreras** [reply]
which versions should I use now? for:

"@walletconnect/core": "2.9.0-e9ace057"
"@walletconnect/types": "2.9.0-e9ace057"
"@walletconnect/web3wallet": "1.8.6-e9ace057"
